### DISCUSSION TITLE
Some methods needed in Framework modules are missing in InnerBrowser

### DISCUSSION THREAD

**Discussion by TavoNiievez** [Codeception/Codeception#6038] [Ideas]
In Symfony, for emails assertions including `seeEmailIsSent` you need to stop the redirects:
 

> If your app performs a redirect after sending the email, you need to suppress this using REST Module's [stopFollowingRedirects](https://codeception.com/docs/modules/REST#stopFollowingRedirects).

So...

#### 2. These two simple call functions are duplicated in both modules...?:

- `setServerParameters()` [[InnerBrowser](https://github.com/Codeception/lib-innerbrowser/blob/fd921e089147057b456ca3660de72112167e40a4/src/Codeception/Lib/InnerBrowser.php#L2047) | [REST](https://github.com/Codeception/module-rest/blob/beeb5a91a97d042273bf10f00063e9b8f541879a/src/Codeception/Module/REST.php#L1617)]
- `haveServerParameter()` [[InnerBrowser](https://github.com/Codeception/lib-innerbrowser/blob/fd921e089147057b456ca3660de72112167e40a4/src/Codeception/Lib/InnerBrowser.php#L2061) | [REST](https://github.com/Codeception/module-rest/blob/beeb5a91a97d042273bf10f00063e9b8f541879a/src/Codeception/Module/REST.php#L1629)]

#### 3. The following methods are relevant to Framework modules and are not in InnerBrowser:

- [x] **followRedirect()**

 If use `stopFollowingRedirects` to stop the redirect, i don't have available the function that allows me to fire that redirect after executing my assertions. `startFollowingRedirects` simply re-enables the next redirect that <b>might occur</b>, but <b>does not execute the pending one</b>. This function is performed by `followRedirect`.
- [ ] **foward()**
 Codeception offers a method to go to the previous page 'moveBack()'... But then it is not possible for me to go forward in the history.
- [ ] **reload()**
 See related issue: https://github.com/Codeception/lib-innerbrowser/issues/33
- [ ] **restart()**
- [x] **setMaxRedirects()**

#### 4. Which ones are really missing and which ones can be contributed?

Thanks in advance.

**Comment by Naktibalda** [comment]
1. Probably because somebody wanted this functionality in REST module. Look at Blame.
2. Let me guess. Somebody decided 3 years later, that another name is more appropriate.
3. I don't remember any intention to expose methods of AbstractBrowser.
4. So none of them are missing.

**Comment by TavoNiievez** [comment]
> 4. So none of them are missing.

So how do i use `->followRedirect()` to execute pending redirects in my tests?
```php
	$I->amOnPage('/register');
	$I->stopFollowingRedirects();
	$I->submitForm('form=name[register_form]', [
		'register_form[email]' => 'john_doe@gmail.com',
		'register_form[password]' => '123456'
	]);
	$I->seeEmailIsSent(1);
	// I can't make the following assertions because i stopped the redirect.
	$I->expect('to be redirected to the login page');
	$I->seeResponseCodeIsSuccessful();
	$I->seeCurrentUrlEquals('/login');
	$I->see('Login');
```
`->startFollowingRedirects()` just enable the next redirect that may occur.

**Comment by Naktibalda** [comment]
Well, nobody asked for that until now.

followRedirects doesn't make much sense in context of REST module and I wasn't aware that somebody is using stopRedirects for Symfony email assertions.

**Comment by Naktibalda** [comment]
I wrote that start/stop FollowingRedirects should be in InnerBrowser 5 years ago: https://github.com/Codeception/Codeception/pull/2598#issuecomment-163905864

**Comment by DavertMik** [comment]
> 1. Why are these functions specific to the REST module?:

For PhpBrowser and Frameworks module I don't assume that someone needs to stop redirects. That's why. In REST you may have more control over requests.

Yeah, I'm also not so sure that this is the huge issue we should discuss. But if you know a better idea on how to restructurize the code , I'm open to it.

**Comment by TavoNiievez** [comment]
The easiest way would be to make [`getRunningClient()`](https://github.com/Codeception/lib-innerbrowser/blob/fd921e089147057b456ca3660de72112167e40a4/src/Codeception/Lib/InnerBrowser.php#L298) `public` in InnerBrowser, 

which would allow anyone to execute the non-wrapper methods that they need...

The other solution would involve creating wrappers for `followRedirect()`, `moveFoward()`, or any other method of item 3 that is considered relevant for testing in the Framework modules.

That would be enough to close this discussion.

**Comment by TavoNiievez** [comment]
See https://github.com/Codeception/lib-innerbrowser/pull/29 .
